### PR TITLE
NerlGUI and py merger

### PR DESCRIPTION
**Pull request by halfway258**
This pull request will close NerlGUI for now and allow to focus on bugs and general ease-of-use of NerlNet

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -59,9 +59,19 @@ After running the installation script, NErlNet directory can be accessed via the
 ## Python API and Jupyter: 
 
 1. Create and activate a virtual environment for Nerlnet https://docs.python.org/3/library/venv.html
-2. Install required modules ```pip install -r src_py/requirements.txt```
+2. Install required modules ```pip install -r src_py/reqCleaned.txt```
```
Comment (leondavi) [inline README.md:62]: @halfway258 
Don't change the name of requirements file 
requirements.txt name is counted as a standard

File: README.md [hunk 2]
```diff
@@ -59,9 +59,19 @@ After running the installation script, NErlNet directory can be accessed via the
...
 3. Install Jupyter Notebook ```pip install jupyterlab```
 4. import the ApiServer module to the notebook. 
 5. Follow the example: https://github.com/leondavi/NErlNet/blob/master/src_py/Deprecated/flow_example.ipynb 
 
+## Running remotely on azure
```
Comment (leondavi) [inline README.md:67]: Remove all the part of running remotely on azure from the readme. 
This is excellent that you wrote it. 
Send me a word document of it. This is only for personal use.

File: inputJsonFiles/Architecture/jsonExampleArch4Devs4Workers1SourceMonitored.json [hunk 3]
```diff
@@ -0,0 +1,177 @@
+{
```
Comment (leondavi) [inline inputJsonFiles/Architecture/jsonExampleArch4Devs4Workers1SourceMonitored.json:1]: Rename this file to adhere the new style of json files

File: inputJsonFiles/Architecture/jsonExampleArch4Devs4Workers2SourcesAzure.json [hunk 4]
```diff
@@ -0,0 +1,179 @@
+{
```
Comment (leondavi) [inline inputJsonFiles/Architecture/jsonExampleArch4Devs4Workers2SourcesAzure.json:1]: Rename this file to adhere the new style of json files

File: inputJsonFiles/ConnectionMap/1Router4Clients2Sources.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (leondavi) [inline inputJsonFiles/ConnectionMap/1Router4Clients2Sources.json:1]: Rename this file to adhere the new style of json files

File: inputJsonFiles/experimentsFlow/jsonExampleExpFlow2Worker2Router1Source.json [hunk 6]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (leondavi) [inline inputJsonFiles/experimentsFlow/jsonExampleExpFlow2Worker2Router1Source.json:1]: Rename this file to adhere the new style of json files

File: jsonPath [hunk 7]
```diff
@@ -1,6 +1,6 @@
-../../../inputJsonFiles/Architecture/jsonArch3PC2Workers1SourceMonitored.json
```
Comment (leondavi) [inline jsonPath:1]: Do we still need it after the change of python we did?

File: src_erl/Communication_Layer/http_Nerlserver/src/MainServer/actionHandler.erl [hunk 8]
```diff
@@ -19,13 +19,13 @@ init(Req0, [Action, Main_genserver_Pid]) ->
...
 
   case Action of
-    clientsTraining ->  gen_statem:cast(Main_genserver_Pid, {clientsTraining,Body});
-    clientsPredict ->  gen_statem:cast(Main_genserver_Pid, {clientsPredict,Body});
-    lossFunction ->  gen_statem:cast(Main_genserver_Pid, {lossFunction,Body});
-    predictRes ->  gen_statem:cast(Main_genserver_Pid, {predictRes,Body});
-    statistics ->  gen_statem:cast(Main_genserver_Pid, {statistics,Body});
-    startCasting ->  gen_statem:cast(Main_genserver_Pid, {startCasting,Body});
-    stopCasting ->  gen_statem:cast(Main_genserver_Pid, {stopCasting,Body})
+    clientsTraining ->  gen_server:cast(Main_genserver_Pid, {clientsTraining,Body});
```
Comment (leondavi) [inline src_erl/Communication_Layer/http_Nerlserver/src/MainServer/actionHandler.erl:22]: @halfway258 @bsyehuda 
Why did you change from statem to server?

File: src_py/apiServer/apiServer.py [hunk 9]
```diff
@@ -43,13 +58,21 @@ def __init__(self):
...
         # Send the content of jsonPath to each devices:
         print("\nSending JSON paths to devices...")
 
-        archAddress = globe.content[0][:-1]
-        connMapAddress = globe.content[1][:-1]
+        # archAddress = globe.content[0][:-1]
+        # connMapAddress = globe.content[1][:-1]
+
+        archAddress , connMapAddress, exp_flow_json = self.getUserJsons()
+        [bad, archPath] = archAddress.split("/NErlNet")
```
Comment (leondavi) [inline src_py/apiServer/apiServer.py:70]: What is bad in here? 
bad is a bad name for a variable 😃 please change it to a more informative name. 
There is also a hidden bug here. If  "/NErlNet" appears more than once in path?

File: src_py/apiServer/apiServer.py [hunk 10]
```diff
@@ -43,13 +58,21 @@ def __init__(self):
...
+        archAddress = "../../.."+archPath
```
Comment (leondavi) [inline src_py/apiServer/apiServer.py:71]: What is the hidden assumption of going 3 directories up?
I don't like this approach.
Please put an explanation here or get rid of these 3 ".."

File: src_py/apiServer/experiment.py [hunk 11]
```diff
@@ -9,20 +9,23 @@
...
         self.name = expName
         self.trainingResList = []
         self.predictionResList = []
+    
+    def set_experiment_flow(self,expFlow):
+        self.expFlow = expFlow
 
     def syncTrainingWithFlow(self):
-       for source in globe.expFlow["Training"]:
+       for source in self.expFlow ["Training"]:
```
Comment (leondavi) [inline src_py/apiServer/experiment.py:21]: Add the global var a definition (like in cpp): 
TRAINING_STR = "Training"
PREDICTION_STR = "Prediction" 
and use the variable. 

Think about the case that we want to change Prediciton to inference in the future (for example).

File: src_py/apiServer/transmitter.py [hunk 12]
```diff
@@ -53,9 +53,9 @@ def startCasting(self, phase): # numOfBatches, is no. of batches to request from
         print('\nStart Casting Phase')
 
         if (phase=="Training"):
-            batchesPerSource = globe.expFlow["Batches per source"]["Training"]
+            batchesPerSource = globe.experiment_flow_global.expFlow["Batches per source"]["Training"]
```
Comment (leondavi) [inline src_py/apiServer/transmitter.py:56]: Add definition in global to "Batches per source as well" 
E.g. 
BATHCHES_PER_SOURCE_STR = "Batches per source

File: src_py/apiServer/transmitter.py [hunk 13]
```diff
@@ -125,8 +125,8 @@ def predict(self):
             time.sleep(0.005)
             pass 
         
-        globe.expResults.remove0Tails()
-        globe.multiProcQueue.put(globe.expResults)
+        globe.experiment_flow_global.remove0Tails()
```
Comment (leondavi) [inline src_py/apiServer/transmitter.py:128]: remove0Tails? is it a typo? 
Why 0?

**Review by leondavi — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (halfway258) [reply README.md:62]: Fixed

File: README.md (hunk 2 above)
Comment (halfway258) [reply README.md]: Fixed

File: src_py/apiServer/apiServer.py (hunk 9 above)
Comment (halfway258) [reply src_py/apiServer/apiServer.py]: Var named changed to JsonsPath.
Jsons are always saved in "/NErlNet/inputJsonFiles/{JSON_TYPE}/file.json", so there isn't another NErlNet dir on that path.

File: src_py/apiServer/apiServer.py (hunk 10 above)
Comment (halfway258) [reply src_py/apiServer/apiServer.py:72]: added explanation

File: src_py/apiServer/experiment.py (hunk 11 above)
Comment (halfway258) [reply src_py/apiServer/experiment.py]: Fixed

File: src_py/apiServer/transmitter.py (hunk 13 above)
Comment (halfway258) [reply src_py/apiServer/transmitter.py:128]: Function created by Dor, by the name I assume it removes trailing "0" from experiment data?

File: jsonPath (hunk 7 above)
Comment (halfway258) [reply jsonPath:1]: Removed.

File: src_py/apiServer/transmitter.py (hunk 13 above)
Comment (leondavi) [reply src_py/apiServer/transmitter.py:128]: Ok let's keep it as it is

**Review by leondavi — APPROVED** [review]
